File: src_mcp/runners/Processes/ProcessLauncher.cs
using System.Diagnostics;
using System.Text;

namespace CoaiMcp.Runners.Processes;

public sealed record ProcessRequest(
    string Executable,
    IReadOnlyList<string> Arguments,
    string WorkingDirectory)
{
    public IReadOnlyDictionary<string, string?> Environment { get; init; } =
        new Dictionary<string, string?>();

    /// <summary>
    /// Written to the child's stdin, then closed. Empty means "close immediately".
    /// </summary>
    /// <remarks>
    /// This is how every long or multi-line input travels. On Windows the vendor CLIs are npm
    /// <c>.cmd</c> shims, which cmd.exe parses — and cmd.exe truncates an argument at its first
    /// newline. The first real run (2026-08-31) passed a full review prompt in argv and the model
    /// answered "No implementation plan was provided": it had received the first line and nothing
    /// else, silently. stdin has no such parser between us and the process.
    /// </remarks>
    public string StdIn { get; init; } = string.Empty;

    public TimeSpan Timeout { get; init; } = TimeSpan.FromMinutes(10);

    /// <summary>
    /// How much of each stream is kept. Past it the rest is dropped and the text says so.
    /// </summary>
    /// <remarks>
    /// <para>CHARACTERS rather than bytes, and the difference is not pedantry: what is bounded is the
    /// memory of a .NET string, which is two bytes per character. The default is therefore
    /// 8 Mi characters — <b>16 MiB</b> per stream, which is the number the plan named.</para>
    /// <para>It has to be enforced while the stream is READ, not per line. A child can write two
    /// hundred megabytes without a single newline; a line-based reader delivers none of it until the
    /// stream closes, so a ceiling checked in a line callback fires only after the allocation it
    /// exists to prevent has already happened. (codex, plan round.)</para>
    /// <para>The child is not killed for reaching it — the rest is read and discarded, so a runaway
    /// cannot block on a pipe nobody is draining either. Its exit code and the head of what it said
    /// are still returned.</para>
    /// </remarks>
    public int MaxOutputChars { get; init; } = 8 * 1024 * 1024;

    /// <summary>
    /// A name to record this child under while it runs, or empty to track nothing.
    /// </summary>
    /// <remarks>
    /// Opt-in per request rather than for everything, because most processes here are git commands
    /// that finish in milliseconds and would only add write traffic. The ones worth tracking are
    /// the reviewers: minutes long, expensive, and the ones that get orphaned.
    /// </remarks>
    public string TrackAs { get; init; } = string.Empty;
}

/// <summary>What a run produced. <see cref="TimedOut"/> true means the tree was killed.</summary>
public sealed record ProcessResult(int ExitCode, string StdOut, string StdErr, bool TimedOut);

/// <summary>The seam every process in this repository goes through — one launcher, injectable
/// everywhere, so a test hands in a fake and the suite touches no vendor.</summary>
public interface IProcessLauncher
{
    Task<ProcessResult> RunAsync(ProcessRequest request, CancellationToken ct = default);
}

/// <summary>
/// The real launcher. On timeout the WHOLE process tree is killed — a reviewer CLI spawns its own
/// children, and an orphaned child that keeps running is a reviewer that never really failed.
/// </summary>
/// <param name="tracker">
/// Where a long-running child is written down while it runs, so a later server can collect it if
/// this one dies. Defaults to recording nothing, which is what every test and every git command
/// wants.
/// </param>
public sealed class ProcessLauncher(IProcessTracker? tracker = null) : IProcessLauncher
{
    private readonly IProcessTracker _tracker = tracker ?? NoProcessTracking.Instance;

    public async Task<ProcessResult> RunAsync(ProcessRequest request, CancellationToken ct = default)
    {
        // Resolved here, at the one place a process is started, so every caller — reviewers, git,
        // the creds probe — gets the npm-shim fix without knowing it exists.
        var info = new ProcessStartInfo(ExecutableResolver.Resolve(request.Executable))
        {
            WorkingDirectory = request.WorkingDirectory,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            RedirectStandardInput = true,
            UseShellExecute = false,
            // UTF-8 on every stream, explicitly. Without this .NET decodes a child's output with
            // the console's ANSI code page, and on Windows that turns every non-ASCII character
            // into '?' — silently. Found by a translation test whose Ukrainian came back as
            // "??? ?????", but it was never only translation: a finding written in any language
            // but English was being corrupted on its way in.
            StandardOutputEncoding = System.Text.Encoding.UTF8,
            StandardErrorEncoding = System.Text.Encoding.UTF8,
            // UTF-8 WITHOUT the byte-order mark, and the difference is not cosmetic.
            // `Encoding.UTF8` carries a preamble, and .NET flushes it into the child's stdin
            // inside Process.Start() — which means every prompt this product has ever sent began
            // with three stray bytes, and a child that exited quickly (git, mostly) took the whole
            // launch down with a broken pipe from INSIDE Start. Found in WSL, where the race is
            // lost reliably rather than once in a hundred runs: five tests, all in Process.Start.
            StandardInputEncoding = new System.Text.UTF8Encoding(encoderShouldEmitUTF8Identifier: false),
        };
        foreach (var argument in request.Arguments)
        {
            info.ArgumentList.Add(argument);
        }

        foreach (var (name, value) in request.Environment)
        {
            info.Environment[name] = value;
        }

        using var process = new Process { StartInfo = info };
        var stdout = new BoundedText(request.MaxOutputChars);
        var stderr = new BoundedText(request.MaxOutputChars);

        process.Start();

        // Recorded as soon as it exists and forgotten however this method leaves — returned,
        // timed out, cancelled or thrown. The record is what a LATER server reads: the kill below
        // is performed by this process, so it cannot happen if this process is the one that dies,
        // which is exactly how a reviewer ends up outliving its round by ten hours.
        var tracked = request.TrackAs.Length > 0;
        if (tracked)
        {
            _tracker.Record(process.Id, process.StartTime.ToUniversalTime(), request.TrackAs);
        }

        try
        {
            return await RunToCompletionAsync(process, request, stdout, stderr, ct);
        }
        finally
        {
            if (tracked)
            {
                _tracker.Forget(process.Id);
            }
        }
    }

    /// <summary>
    /// One deadline over the WHOLE operation — the write, the wait and the reading alike.
    /// </summary>
    /// <remarks>
    /// <para><b>The write used to happen before the clock started, and that was the defect.</b> A
    /// pipe holds about 4 KiB on Windows and 64 KiB on Linux; a review prompt is a shaped diff of up
    /// to 192 KiB, so every reviewer launch writes past the buffer and blocks until the child reads.
    /// A child that never reads — a sign-in prompt, a TTY check, a CLI that died before its first
    /// read — held the launch for as long as it lived, with no timeout started and the caller's
    /// token unconsulted. Measured 2026-09-09: a ten-second child, a 300 ms budget, and a call that
    /// returned after ten seconds saying <c>TimedOut=false</c>, because by the time the clock was
    /// created the process had already gone.</para>
    /// <para>So everything below runs under ONE linked, timed token created before a byte is
    /// written, and the three concurrent halves — writing, reading, waiting — are all ended by it.
    /// The kill is what actually frees a blocked write: it closes the child's end of the pipe, so a
    /// write that ignored its token (an anonymous pipe on Windows does) still fails at once, with
    /// the <c>IOException</c> this class has always treated as the child's decision.</para>
    /// </remarks>
    private static async Task<ProcessResult> RunToCompletionAsync(
        Process process,
        ProcessRequest request,
        BoundedText stdout,
        BoundedText stderr,
        CancellationToken ct)
    {
        using var deadline = CancellationTokenSource.CreateLinkedTokenSource(ct);
        deadline.CancelAfter(request.Timeout);

        // Started before the write, because a child may answer before it has finished reading — and
        // one that fills ITS pipe while nobody drains ours is a deadlock made of two blocked writes.
        var reading = Task.WhenAll(
            DrainAsync(process.StandardOutput, stdout),
            DrainAsync(process.StandardError, stderr));
        var writing = WriteStdInAsync(process, request.StdIn, deadline.Token);

        var timedOut = false;
        try
        {
            await process.WaitForExitAsync(deadline.Token);
        }
        catch (OperationCanceledException)
        {
            timedOut = true;
            try
            {
                process.Kill(entireProcessTree: true);
            }
            catch (InvalidOperationException)
            {
                // It exited between the deadline and the kill; the race loser is fine either way.
            }

            await process.WaitForExitAsync(CancellationToken.None);
        }

        await SettledAsync(writing);
        await SettledAsync(DrainedAsync(reading));

        return new ProcessResult(process.ExitCode, stdout.ToString(), stderr.ToString(), timedOut);
    }

    /// <summary>Reads one stream to its end, keeping at most what <paramref name="into"/> allows.</summary>
    /// <remarks>
    /// Character by chunk rather than line by line. The line-based reader this replaced could not
    /// enforce a ceiling at all — it delivers nothing until a newline arrives, so a child writing two
    /// hundred megabytes without one had already been buffered by the framework before any callback
    /// ran. It also APPENDED a newline to every line, so a vendor's answer came back with a line
    /// ending it had not written; the raw reader hands over exactly what the child produced.
    /// </remarks>
    private static async Task DrainAsync(StreamReader stream, BoundedText into)
    {
        var buffer = new char[8192];
        while (await stream.ReadAsync(buffer, CancellationToken.None) is var read && read > 0)
        {
            into.Append(buffer, read);
        }
    }

    /// <summary>
    /// Waits for both streams to end, but not for ever.
    /// </summary>
    /// <remarks>
    /// A pipe stays open while ANY process holds its write end, and a grandchild inherits both. The
    /// tree kill takes them, so this normally returns the instant the child is gone — but a handle
    /// that outlives the kill would otherwise hang this method rather than the process it belongs
    /// to, and a launcher that never returns is worse than a truncated stream.
    /// </remarks>
    private static async Task DrainedAsync(Task reading)
    {
        if (await Task.WhenAny(reading, Task.Delay(TimeSpan.FromSeconds(5))) == reading)
        {
            await reading;
        }
    }

    /// <summary>
    /// Awaits a half of the operation, treating the child's own departure as an ordinary end.
    /// </summary>
    /// <remarks>
    /// These three are what a dead child leaves behind: a broken pipe, a stream disposed under a
    /// write in flight, and the deadline that ended the write itself. None of them is a failure of
    /// the LAUNCH — the child's exit code and its own words are the answer — and letting one travel
    /// out of here failed a whole round rather than one reviewer with a named outcome, which is a
    /// flake this class has already paid for once.
    /// </remarks>
    private static async Task SettledAsync(Task half)
    {
        try
        {
            await half;
        }
        catch (Exception e) when (e is IOException or ObjectDisposedException or OperationCanceledException)
        {
        }
    }

    /// <summary>
    /// Feeds the child its input, and treats a closed pipe as the child's decision rather than
    /// our failure.
    /// </summary>
    /// <remarks>
    /// A process that exits before reading — a git command that needed no input, a CLI that
    /// refused at once, one that read the first line and stopped — closes its end of the pipe, and
    /// the write then throws. Left unguarded that exception came out of the launcher instead of a
    /// <see cref="ProcessResult"/>, so an early exit failed the whole ROUND rather than one
    /// reviewer with a named outcome. It surfaced as a rare cross-platform test flake for exactly
    /// as long as it took the race to be lost twice.
    /// <para>Nothing is hidden: the child's exit code, stdout and stderr are still returned, so a
    /// CLI that exited early because it was broken is still reported as broken.</para>
    /// <para>The CLOSE is what a child waiting for EOF is waiting for, so it happens on every path
    /// including a write the deadline ended — otherwise a child that read everything would sit
    /// waiting for an end-of-file that never came.</para>
    /// </remarks>
    private static async Task WriteStdInAsync(Process process, string text, CancellationToken ct)
    {
        try
        {
            if (text.Length > 0)
            {
                await process.StandardInput.WriteAsync(text.AsMemory(), ct);
            }
        }
        finally
        {
            try
            {
                process.StandardInput.Close();
            }
            catch (Exception e) when (e is IOException or ObjectDisposedException)
            {
                // The stream is already gone with the child. There is nothing left to close.
            }
        }
    }
}

/// <summary>
/// One stream's text, up to a ceiling, with the truncation said out loud.
/// </summary>
/// <remarks>
/// Not a <see cref="StringBuilder"/> with a check beside it, because the check has to hold at every
/// call site and there are two of them per launch. A bound that lives in the type cannot be
/// forgotten by the next reader of this file.
/// </remarks>
internal sealed class BoundedText(int maxChars)
{
    private readonly StringBuilder _text = new();
    private bool _truncated;

    public void Append(char[] buffer, int count)
    {
        if (_truncated)
        {
            // Still READ — the caller keeps draining so the child never blocks on a pipe nobody
            // empties — and no longer kept.
            return;
        }

        var room = maxChars - _text.Length;
        if (count <= room)
        {
            _text.Append(buffer, 0, count);

            return;
        }

        _text.Append(buffer, 0, Math.Max(room, 0));
        // Named, once, in the stream itself. A silently cut answer is one a parser fails on for a
        // reason nobody can find; this one fails saying why.
        _text.Append($"\n[coai: output truncated after {maxChars} characters]\n");
        _truncated = true;
    }

    public override string ToString() => _text.ToString();
}

File: src_vs_code/src/processLauncher.ts
import { spawn } from 'node:child_process';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

/**
 * The one place this extension starts a process it controls.
 *
 * <p>It is the version probe's own launcher, WIDENED rather than copied. Every guard below was
 * written against a real failure — a `.cmd` that throws synchronously instead of emitting `error`,
 * a shell probe whose grandchild outlived the kill, a bare `taskkill` that could be a file planted
 * in an opened workspace, a working directory `cmd.exe` searches before the PATH. A second launcher
 * would have had to learn each of them again, and the one already in the tree (`wslNetwork.ts`)
 * shows how that goes: it has a different stdio contract and none of this hardening.</p>
 *
 * <p><b>Why a handle and not a promise.</b> `capture` wants the whole output and then the exit code,
 * which a promise says well. A chat session wants to write a line, read lines back, and keep the
 * process for the next question — which a promise cannot say at all. One launcher answers both.</p>
 *
 * <p><b>The one deliberate exception to "one spawn site".</b> `killTree` starts `taskkill.exe`
 * itself, so this file contains two `spawn` calls rather than one. It is named here because a
 * structural guard that counts spawns will find it and should not treat it as a regression: killing
 * a tree is not launching a child, it takes no handle, produces no output anybody reads, and is
 * unref'd immediately. Raised by codex on the code round of this very change, and answered by saying
 * so rather than by hiding it behind an injected primitive that would have one implementation.</p>
 *
 * <p><b>Two details are load-bearing.</b> The exit event fires on `close`, not `exit`: `exit` can
 * arrive with output still buffered, and a truncated banner parses to no version at all. And every
 * subscription is late-safe — output that arrived before anybody subscribed is replayed to the first
 * subscriber, because a short-lived child can print its whole answer and close before the caller's
 * next line runs.</p>
 */

/** How much stderr is worth keeping. Enough for a stack or a refusal; not a leak. */
const STDERR_TAIL_MAX = 8000;

/** Stop receiving. Returned by every stream subscription — see `ProcessHandle.onLine`. */
export type Unsubscribe = () => void;

export interface LaunchOptions {
  /** Run through the platform shell. Only the Windows shim case sets this — see `needsShell`. */
  readonly shell?: boolean;
  /** The child's working directory. Defaults per `workingDirectory` — a shell never gets the cwd. */
  readonly cwd?: string;
}

export interface ProcessHandle {
  /**
   * The child's process id, or 0 when it never started.
   *
   * <p>For the LEDGER, and for nothing else here: a chat child is written down as it starts so a
   * force-killed editor cannot leave an authenticated CLI running. Killing by this number is only
   * safe against a process whose identity has been re-checked — see `chatLedger.ts`, and the note
   * on `killTree` below for why the launcher itself never does it.</p>
   */
  readonly pid: number;
  /** Write one line (a newline is appended). `false` when the pipe is already gone. */
  writeLine(line: string): boolean;
  /**
   * Write, then CLOSE the input — for a child that reads its whole prompt and then works.
   *
   * <p>`codex exec -` is the case: it reads instructions from stdin and waits for more until the
   * stream ends, so a turn written without this simply never starts. A persistent vendor must never
   * be sent it — closing the pipe there ends the conversation.</p>
   */
  writeAndEnd(text: string): boolean;
  /** Raw stdout as it arrives. Replayed to the first subscriber; unsubscribe when done. */
  onStdout(listener: (chunk: string) => void): Unsubscribe;
  /** Whole stdout lines, split across chunk boundaries, each delivered once. */
  onLine(listener: (line: string) => void): Unsubscribe;
  /** The exit code, once, on `close`. Never fires after `onError`. Late-safe. */
  onExit(listener: (code: number) => void): void;
  /** The child could not be started, or died of its own error. Fires once. Late-safe. */
  onError(listener: (reason: string) => void): void;
  /** The tail of what the child wrote to stderr — or why it never started. */
  stderrTail(): string;
  /** Kill what we started — the whole tree when a shell is in the way. Safe to call twice. */
  kill(): void;
}

/**
 * Where a child runs.
 *
 * <p>Its own function so the rule can be TESTED rather than trusted: a shell that is handed no
 * working directory inherits the extension host's, which in VS Code is the opened workspace — and
 * `cmd.exe` searches the working directory before the PATH. A workspace holding a file named like
 * the tool being probed would then run with the host's privileges. The caller may still choose a
 * directory; what it may not do is leave a shell pointed at somebody's checkout by omission.
 * (gemini, the code round.)</p>
 */
export function workingDirectory(shell: boolean, cwd: string | undefined): string | undefined {
  if (cwd !== undefined) {
    return cwd;
  }

  return shell ? tmpdir() : undefined;
}

/**
 * Start a process, or answer with a handle that is already in the error state.
 *
 * <p><b>`spawn` can THROW rather than emit `error`</b>, and it does for a real case: node refuses a
 * `.cmd` without a shell with a synchronous `EINVAL` (the 2024 argument-injection fix). Measured on
 * `codex.cmd`, where the exception escaped `render` and the panel stopped repainting — which is why
 * the try/catch is around the call itself and not only in an `error` handler.</p>
 */
export function launch(target: string, args: readonly string[], options: LaunchOptions = {}): ProcessHandle {
  const shell = options.shell === true;
  const cwd = workingDirectory(shell, options.cwd);

  let child: ReturnType<typeof spawn> | undefined;
  let failure = '';
  try {
    child = spawn(target, [...args], {
      shell,
      windowsHide: true,
      ...(cwd === undefined ? {} : { cwd }),
    });
  } catch (reason) {
    failure = reason instanceof Error ? reason.message : String(reason);
  }

  return child === undefined ? failedHandle(failure) : liveHandle(child, shell);
}

/**
 * A handle for a child that never started.
 *
 * <p>Its `stderrTail` carries the REASON rather than an empty string. A caller that shows "could not
 * be read" and nothing else leaves somebody guessing between a missing binary, a denied permission
 * and a rejected argument — three different things to do next. (local and gemini, the code round.)</p>
 */
function failedHandle(reason: string): ProcessHandle {
  return {
    pid: 0,
    writeLine: () => false,
    writeAndEnd: () => false,
    onStdout: () => () => undefined,
    onLine: () => () => undefined,
    onExit: () => undefined,
    onError: (listener) => {
      listener(reason);
    },
    stderrTail: () => reason,
    kill: () => undefined,
  };
}

/**
 * A stream of values that holds what it emitted until somebody is listening.
 *
 * <p>The replay is the point. A child can write its whole answer and close before the caller's next
 * statement runs, and an `onLine` that only appends a listener loses that answer for ever — codex
 * and gemini raised it independently. What is held is handed to the FIRST subscriber and then
 * dropped: a second subscriber joining later is joining a conversation in progress, and pretending
 * otherwise would deliver one line twice.</p>
 */
function replayingFan<T>(): { emit(value: T): void; on(listener: (value: T) => void): Unsubscribe } {
  let listeners: readonly ((value: T) => void)[] = [];
  let held: readonly T[] = [];
  let opened = false;

  return {
    emit(value: T): void {
      if (!opened) {
        held = [...held, value];
        return;
      }
      for (const listener of listeners) {
        listener(value);
      }
    },
    on(listener: (value: T) => void): Unsubscribe {
      opened = true;
      listeners = [...listeners, listener];
      const pending = held;
      held = [];
      for (const value of pending) {
        listener(value);
      }

      // Without this, a chat session that subscribes per turn leaves every past turn's callback in
      // the list, and turn 100's answer is delivered to ninety-nine stale handlers. (codex, gemini.)
      return () => {
        listeners = listeners.filter((known) => known !== listener);
      };
    },
  };
}

/** A thing that happens at most once, and is still heard by whoever asks afterwards. */
function onceFan<T>(): { fire(value: T): boolean; on(listener: (value: T) => void): void } {
  let listeners: readonly ((value: T) => void)[] = [];
  let fired: { value: T } | undefined;

  return {
    fire(value: T): boolean {
      if (fired !== undefined) {
        return false;
      }
      fired = { value };
      for (const listener of listeners) {
        listener(value);
      }

      return true;
    },
    on(listener: (value: T) => void): void {
      if (fired !== undefined) {
        listener(fired.value);
        return;
      }
      listeners = [...listeners, listener];
    },
  };
}

/**
 * Whole lines out of a stream that arrives in arbitrary pieces.
 *
 * <p>A line is only whole once its terminator has arrived; whatever follows the last one waits for
 * the next chunk. This is the split a naive `chunk.split('\n')` gets wrong exactly when the pipe is
 * busiest. `flush` exists because a child that prints without a final newline still printed a line.</p>
 */
function lineSplitter(deliver: (line: string) => void): { push(text: string): void; flush(): void } {
  let pending = '';

  return {
    push(text: string): void {
      pending += text;
      const parts = pending.split(/\r?\n/);
      pending = parts.pop() ?? '';
      for (const line of parts) {
        deliver(line);
      }
    },
    flush(): void {
      if (pending.length === 0) {
        return;
      }
      const last = pending;
      pending = '';
      deliver(last);
    },
  };
}

function liveHandle(child: ReturnType<typeof spawn>, shell: boolean): ProcessHandle {
  const stdout = replayingFan<string>();
  const lines = replayingFan<string>();
  const exited = onceFan<number>();
  const errored = onceFan<string>();
  const splitter = lineSplitter((line) => lines.emit(line));
  const stderr = { tail: '' };

  // Decode as text on the STREAM, not per chunk: a Cyrillic character split across a chunk boundary
  // becomes two replacement characters if each half is decoded alone, and the passages this launcher
  // exists to carry are routinely Russian. (gemini, the code round.)
  child.stdout?.setEncoding('utf8');
  child.stderr?.setEncoding('utf8');

  child.stdout?.on('data', (text: string) => {
    stdout.emit(text);
    splitter.push(text);
  });
  child.stderr?.on('data', (text: string) => {
    stderr.tail = `${stderr.tail}${text}`.slice(-STDERR_TAIL_MAX);
  });

  // EPIPE on a pipe whose reader has gone arrives ASYNCHRONOUSLY, so no try/catch around `write` can
  // see it. Unhandled, node makes it a fatal exception and the extension host dies with it — the one
  // finding in this round that took a whole editor down. (gemini, Blocking.)
  child.stdin?.on('error', () => undefined);

  child.on('error', (reason: Error) => {
    // The reason goes into the tail as well as to the listener. A missing binary does NOT throw
    // synchronously on Windows — it arrives here, asynchronously — so a caller that only reads
    // `stderrTail` to explain a failure would have found it empty for the commonest failure there
    // is. Caught by the test written for the failed-handle half of the same finding.
    stderr.tail = `${stderr.tail}[coai] ${reason.message}`.slice(-STDERR_TAIL_MAX);
    errored.fire(reason.message);
  });
  child.on('close', (code) => {
    splitter.flush();
    exited.fire(code ?? -1);
  });

  return {
    pid: child.pid ?? 0,
    writeLine: (line) => writeLine(child, line),
    writeAndEnd: (text) => writeAndEnd(child, text),
    onStdout: stdout.on,
    onLine: lines.on,
    onExit: exited.on,
    onError: errored.on,
    stderrTail: () => stderr.tail,
    kill: () => {
      const failed = killTree(child, shell);
      if (failed.length > 0) {
        // Not swallowed: a tree kill that could not even start is why a vendor process is still
        // running, and that sentence belongs where the caller already looks. (codex, the code round.)
        stderr.tail = `${stderr.tail}\n[coai] tree kill failed: ${failed}`.slice(-STDERR_TAIL_MAX);
      }
    },
  };
}

/** The whole prompt and then EOF, or `false` when the pipe has gone. Never throws. */
function writeAndEnd(child: ReturnType<typeof spawn>, text: string): boolean {
  const stdin = child.stdin;
  if (stdin === null || stdin.destroyed || stdin.writableEnded) {
    return false;
  }
  try {
    stdin.end(text);

    return true;
  } catch {
    return false;
  }
}

/** One line into the child, or `false` when the pipe has gone. Never throws. */
function writeLine(child: ReturnType<typeof spawn>, line: string): boolean {
  const stdin = child.stdin;
  if (stdin === null || stdin.destroyed || stdin.writableEnded) {
    return false;
  }
  try {
    stdin.write(`${line}\n`);

    return true;
  } catch {
    // A pipe that closed between the check and the write is a state, not an exception: the caller
    // learns the turn failed from `onExit`/`onError`, which is where it can say so.
    return false;
  }
}

/**
 * Kill what we started — the whole tree when a shell is between us and the real process.
 *
 * <p><b>`child.kill()`, never `process.kill(pid)`</b>: a probe that exits in the same tick the timer
 * fires has a pid that no longer exists, and Windows reuses pids — so killing by number can throw
 * `ESRCH` or, worse, terminate whatever now holds that number. `child.kill()` on an exited child is
 * a no-op. (codex, an earlier code round.)</p>
 *
 * <p><b>`taskkill` by ABSOLUTE path.</b> `CreateProcess` searches the application directory and the
 * working directory before the system one, so a bare `taskkill` could be a file planted in an opened
 * workspace, run with the extension host's privileges.</p>
 *
 * @returns why the tree kill could not be started, or an empty string when it was
 */
function killTree(child: ReturnType<typeof spawn>, shell: boolean): string {
  const pid = child.pid;
  if (!shell || process.platform !== 'win32' || pid === undefined) {
    child.kill();

    return '';
  }

  const failed = spawnTaskkill(pid);
  child.kill();

  return failed;
}

/**
 * The second `spawn` in this file, in a function named so an audit can find it.
 *
 * <p>A guard that counts spawn calls will see two here and should see two: killing a tree is not
 * launching a child. It takes no handle, produces no output anybody reads, and is unref'd at once.
 * It is a named function rather than a line inside `killTree` for exactly that reason — the
 * exception should be greppable, not buried. (local, the second code round.)</p>
 *
 * <p>Both options are load-bearing, not copied habit. `windowsHide` keeps a console window from
 * flashing over the editor every time a probe times out — eight seconds apart, for as long as the
 * panel repaints. `cwd` is the temp directory for the same reason every shell launch gets one:
 * `CreateProcess` searches the working directory before the system one.</p>
 *
 * <p>`.on('error')` comes BEFORE `.unref()`: a taskkill that fails to spawn emits `error` on a child
 * nobody is listening to, and node turns that into a fatal exception in the extension host.
 * (gemini, the first code round.)</p>
 *
 * @returns why it could not be started, or an empty string
 */
function spawnTaskkill(pid: number): string {
  let failed = '';
  try {
    spawn(TASKKILL, ['/pid', String(pid), '/t', '/f'], { windowsHide: true, cwd: tmpdir() })
      .on('error', (reason: Error) => {
        failed = reason.message;
      })
      .unref();
  } catch (reason) {
    failed = reason instanceof Error ? reason.message : String(reason);
  }

  return failed;
}

/**
 * The system utility, not whatever is called that on the PATH or in a workspace.
 *
 * <p>`WINDIR` before the literal: a Windows installed on `D:` with `SystemRoot` stripped from a
 * custom child environment would otherwise point the tree kill at a path that does not exist, and
 * the failure would be silent apart from a line in the stderr tail. (gemini, the second code round.)</p>
 */
const TASKKILL = join(
  process.env['SystemRoot'] ?? process.env['WINDIR'] ?? 'C:\\Windows',
  'System32',
  'taskkill.exe',
);
